message_ix/tests/test_feature_temporal_level.py

```diff
@@ -0,0 +1,568 @@
+                    yr,
+                    "mode",
+                    "node",
+                    com_dict[tec]["input"],
+                    "level",
+                    h1,
+                    h2,
+                ]
+                scen.add_par("input", ["node", tec] + inp_spec, 1, "-")
+
+    # Adding capacity related parameters
+    if capacity:
+        add_cap_par(scen, [2020], "gas_ppl")
+
+    # Committing and solving
+    scen.commit("scenario was set up.")
+    scen.solve(case=comment)
+
+    # Reading "ACT" greater than zero from the solution
+    act = scen.var("ACT")[scen.var("ACT")["lvl"] > 0]
+
+    # 1) Testing if linkage between "gas_ppl" with "gas_supply" technology is made
+    assert "gas_supply" in set(act["technology"])
+
+    # 2) Testing if "ACT" of "gas_ppl" is correct (with respect to "duration_time_rel")
+    # i.e., sum("ACT" of "gas_ppl") = sum("ACT" of "gas_supply") = sum("demand")
+    assert (
+        sum(act.loc[act["technology"] == "gas_ppl"]["lvl"])
+        == sum(act.loc[act["technology"] == "gas_supply"]["lvl"])
+        == sum(scen.par("demand")["value"])
+    )
+
+    # 3) Test if "CAP" of "gas_ppl" is correctly calculated based on "ACT"
+    # i.e., "CAP" = max("ACT" / "duration_time")
+    if capacity:
+        for h in act.loc[act["technology"] == "gas_ppl"]["time"]:
+            act.loc[
+                (act["time"] == h) & (act["technology"] == "gas_ppl"),
+                "capacity-corrected",
+            ] = act["lvl"] / float(scen.par("duration_time", {"time": h})["value"])
+        assert max(
+            act.loc[act["technology"] == "gas_ppl", "capacity-corrected"]
+        ) == float(scen.var("CAP", {"technology": "gas_ppl"})["lvl"])
+
+
+# Main tests for commodity balance over various temporal levels (and "time" index)
+# In these tests (11 scenarios in total), "demand" is defined in different time slices
+# with different "duration_time", there is one power plant ("gas_ppl") to meet "demand",
+# which receives fuel from a supply technology ("gas_supply").
+# Different temporal level hierarchies are tested, by linkages of "ACT" with
+# "demand" and "CAP".
+
+# 1) "gas_ppl" is active in "summer" and NOT linked to "gas_supply" in "year"
+# This setup should not solve, as no linkgae between power plant and fuel supply.
+def test_temporal_levels_not_linked(test_mp):
+    comment = "1.not-linked"
+    # Dictionary of technology input/output
+    tec_dict = {
+        "gas_ppl": {
+            "time_origin": ["summer"],
+            "time": ["summer"],
+            "time_dest": ["summer"],
+        },
+        "gas_supply": {"time_origin": [], "time": ["year"], "time_dest": ["year"]},
+    }
+
+    # Check the model not solve if there is no link between fuel supply and power plant
+    try:
+        model_generator(
+            test_mp,
+            comment,
+            tec_dict,
+            time_steps=[("summer", 1, "season", "year")],
+            demand={"summer": 2},
+        ),
```

Comment:
Probably don't intend to have a trailing comma here.

Reply:
removed